pkg/reconciler/buildrun/resources/service_accounts_test.go

```diff
@@ -17,45 +19,154 @@ import (
+			Expect(sa.Secrets[0].Name).To(Equal("foosecret"))
+		})
+		It("should return a namespace default sa with a secret reference", func() {
+			buildRunSample := ctl.BuildRunWithoutSA(buildRunName, buildName)
+
+			// stub a GET API call for a service account
+			client.GetCalls(generateGetSAStub("default"))
+
+			sa, err := resources.RetrieveServiceAccount(context.TODO(), client, ctl.BuildWithOutputSecret(buildName, "default", "foosecret"), buildRunSample)
+			Expect(err).To(BeNil())
+			Expect(len(sa.Secrets)).To(Equal(1))
+			Expect(sa.Secrets[0].Name).To(Equal("foosecret"))
+		})
+
+		It("should return an error if the specified sa is not found", func() {
+			buildRunSample := ctl.BuildRunWithSA(buildRunName, buildName, "foobarsa")
+
+			client.GetCalls(generateGetSAStubWithError(k8serrors.NewNotFound(schema.GroupResource{}, "")))
+
+			client.StatusCalls(func() crc.StatusWriter {
+				statusWriter := &fakes.FakeStatusWriter{}
+				statusWriter.UpdateCalls(func(ctx context.Context, object runtime.Object, _ ...crc.UpdateOption) error {
+					return nil
+				})
+				return statusWriter
+			})
+
+			sa, err := resources.RetrieveServiceAccount(context.TODO(), client, ctl.BuildWithOutputSecret(buildName, "default", "foosecret"), buildRunSample)
+			Expect(sa).To(BeNil())
+			Expect(err).ToNot(BeNil())
+			Expect(k8serrors.IsNotFound(err)).To(BeTrue())
+		})
+
+		It("should return multiple errors if the specified sa is not found and the condition update is not working", func() {
+			buildRunSample := ctl.BuildRunWithSA(buildRunName, buildName, "foobarsa")
+
+			client.GetCalls(generateGetSAStubWithError(k8serrors.NewNotFound(schema.GroupResource{}, "")))
+
+			client.StatusCalls(func() crc.StatusWriter {
+				statusWriter := &fakes.FakeStatusWriter{}
+				statusWriter.UpdateCalls(func(ctx context.Context, object runtime.Object, _ ...crc.UpdateOption) error {
+					switch object.(type) {
+					case *buildv1alpha1.BuildRun:
+						return fmt.Errorf("failed")
+					}
+					return nil
+				})
+				return statusWriter
+			})
+
+			sa, err := resources.RetrieveServiceAccount(context.TODO(), client, ctl.BuildWithOutputSecret(buildName, "default", "foosecret"), buildRunSample)
+			Expect(sa).To(BeNil())
+			Expect(err).ToNot(BeNil())
+			Expect(resources.IsClientStatusUpdateError(err)).To(BeTrue())
 		})
+	})
+
+	Context("Retrieving autogenerated service accounts when the spec.serviceAccount.generate key is used", func() {
 
 		It("should provide a generated sa name", func() {
 			Expect(resources.GetGeneratedServiceAccountName(buildRunSample)).To(Equal(buildRunSample.Name + "-sa"))
 		})
+		It("should return a generated sa with a label, ownerreference and a ref secret if it does not exists", func() {
+			buildRunSample := ctl.BuildRunWithSAGenerate(buildRunName, buildName)
+
+			// stub a GET API call for a service account
+			client.GetCalls(generateGetSAStubWithError(k8serrors.NewNotFound(schema.GroupResource{}, "foobar")))
+
+			mountTokenVal := false
```

Comment:
To be able to find the bug that you had with the misplaced `ApplyCredentials`, one can verify that the object is already correct when the `client.Create` call happens (just as suggestion, I know we could be doing this at probably a lot other places as well):

```suggestion
			mountTokenVal := false

			// ensure that the service account is fully populated in the create call
			client.CreateCalls(func(context context.Context, object runtime.Object, _ ...crc.CreateOption) error {
				switch object := object.(type) {
				case *corev1.ServiceAccount:
					Expect(len(object.Secrets)).To(Equal(1))
					Expect(len(object.OwnerReferences)).To(Equal(1))
					Expect(object.Labels[buildv1alpha1.LabelBuildRun]).To(Equal(buildRunName))
					Expect(object.Secrets[0].Name).To(Equal("foosecret"))
					Expect(object.AutomountServiceAccountToken).To(Equal(&mountTokenVal))
				}
				return nil
			})

```

Reply:
added this manually, danke